processor_pool.go

```diff
@@ -107,10 +108,23 @@ func (p *ProcessorPool) Run(poolSize int, queue JobQueue) error {
 
 // GracefulShutdown causes each processor in the pool to start its graceful
 // shutdown.
-func (p *ProcessorPool) GracefulShutdown() {
+func (p *ProcessorPool) GracefulShutdown(pause bool) {
 	p.processorsLock.Lock()
 	defer p.processorsLock.Unlock()
 
+	log := context.LoggerFromContext(p.Context)
+
+	if pause {
+		if p.paused {
+			log.Info("finishing wait group to unpause")
+			p.processorsWG.Done()
+		} else {
+			log.Info("incrementing wait group for pause")
+			p.processorsWG.Add(1)
+			p.paused = true
```

Comment:
This and the read above it should probably be wrapped in a mutex to prevent race conditions. You'd have to send two signals really close to eachother for a race condition to happen, but if it does happen it looks to me like you'd either (1) get a panic because the wait group goes negative or (2) graceful shutdown stops working because each signal "pauses" and adds 1 to the wait group (so +2 in total), but when you unpause, it only removes 1, so you'll always have 1 left in the wait group.